### DISCUSSION TITLE
TinyMCE code plugin removes <a> links

### DISCUSSION THREAD

**Discussion by kchandran-fs** [tinymce/tinymce#9041] [General]
I'm facing an issue with **code** plugin in tinyMCE where I'm pasting an HTML table with <a> tag in the code plugin but when I get back the content from the editor using getContent it removes the <a> tag. 

Here is the content I'm pasting
```
<tr>
<td style="padding:7px 20px 5px 20px" align="left" valign="top" class="box">
<a target="_blank" rel="nofollow" href=https://out.smore.com/e/prvtu/FWGFJL?__$u__ class="skip-tracking"><h3 style="margin:0 0 0 0">
<div style="font-size:28px;text-align:center;line-height:1;" dir="ltr">
<img src=https://cdn.smore.com/e/prvtu/HuPkp93BRz.png alt="Bally " width="62" height="27.93" style="display:inline">
<img src=https://cdn.smore.com/e/prvtu/yDKXyTLXar.png alt="Sports " width="78" height="27.93" style="display:inline">
<img src=https://cdn.smore.com/e/prvtu/dveWvU1Sbp.png alt="Spotlight " width="105" height="27.93" style="display:inline">
<img src=https://cdn.smore.com/e/prvtu/pYshocywUB.png alt="on " width="35" height="27.93" style="display:inline">
<img src=https://cdn.smore.com/e/prvtu/QZO0081Vhk.png alt="Richardson " width="131" height="27.93" style="display:inline">
<img src=https://cdn.smore.com/e/prvtu/OfdhlVycOo.png alt="Softball " width="92" height="27.93" style="display:inline">
</div>
</h3>
</a>
</td>
</tr>

```

Code plugin returned HTML:

```
<tr>\n
<td class="box" style="padding: 7px 20px 5px 20px;" align="left" valign="top">\n
<div dir="ltr" style="font-size: 28px; text-align: center; line-height: 1;">
<img style="display: inline;" src="https://cdn.smore.com/e/prvtu/HuPkp93BRz.png" alt="Bally " width="62" height="27.93" />
<img style="display: inline;" src="https://cdn.smore.com/e/prvtu/yDKXyTLXar.png" alt="Sports " width="78" height="27.93" />
<img style="display: inline;" src="https://cdn.smore.com/e/prvtu/dveWvU1Sbp.png" alt="Spotlight " width="105" height="27.93" />
<img style="display: inline;" src="https://cdn.smore.com/e/prvtu/pYshocywUB.png" alt="on " width="35" height="27.93" />
<img style="display: inline;" src="https://cdn.smore.com/e/prvtu/QZO0081Vhk.png" alt="Richardson " width="131" height="27.93" />
<img style="display: inline;" src="https://cdn.smore.com/e/prvtu/OfdhlVycOo.png" alt="Softball " width="92" height="27.93" />
</div>
\n</td>\n
</tr>

```
TinyMCE version I'm using 4.6.5

**Comment by mrinasugosh** [comment]
The HTML you're pasting into TinyMCE's code plugin is being sanitized/modified when you retrieve it using `getContent`. TinyMCE has an internal logic that cleans up and optimizes the HTML content for consistency and to avoid potential issues. This process includes:

1. Removing potentially unsafe attributes and tags (to prevent XSS attacks).
2. Ensuring that the HTML structure is valid, e.g., not allowing block elements inside inline elements.
3. Formatting and making the HTML more readable.
https://www.tiny.cloud/docs/tinymce/6/security/#sanitizing-html-input-to-protect-against-xss-attacks

From the provided example, TinyMCE has removed the `<a>` and `<h3>` tags and has slightly reformatted the rest of the code. The removal of `<h3>` inside an anchor (`<a>`) is because block-level elements (like `<h3>`) should not be placed inside inline elements (like `<a>`).

To address the behavior and retain the original content:

- **Use valid HTML:** Always try to use valid HTML according to the specifications. This minimizes the risk of TinyMCE reformatting your code. For example, consider placing the anchor tag (`<a>`) inside the header tag (`<h3>`) if possible.
- [**Extended Valid Elements**](https://www.tiny.cloud/docs/tinymce/6/content-filtering/#example-using-extended_valid_elements): You can configure TinyMCE to accept more tags and attributes by using the extended_valid_elements option in the TinyMCE configuration. The extended_valid_elements allows you to specify which tags and their respective attributes should be retained, even if they're not part of the default set considered valid by TinyMCE.
- **Disable Automatic Cleanup**: If you want to completely turn off the automatic cleanup behavior of TinyMCE, you can set the [xss_sanitization](https://www.tiny.cloud/docs/tinymce/6/6.4.1-release-notes/#new-xss_sanitization-option-to-allow-for-xss-sanitization-to-be-disabled) option to false. Turning off the xss_sanitization can make your application vulnerable to XSS attacks if user-generated content is not sanitized elsewhere (NOT RECOMMENDED)

**Reply by redjard** [reply]
Putting non-inline elements inside anchor `<a>` elements is officially supported/in-spec since html5:
> Content model:
>     Transparent, but there must be no interactive content descendant, a element descendant, or descendant with the tabindex attribute specified.
- https://html.spec.whatwg.org/#the-a-element

This excludes elements like `<input>` or `<iframe>`, but allows `<h3>`, `<div>`, `<p>`, ... which are all removed by tinymce.
It would be great if you could follow the official html5 spec and common usage here instead of html4.

Making an entire block a link is impossible using only inline-elements, it would lead to gaps in the clickable area, annoying users.
